### PR TITLE
Add `esp32_camera_web_server:` to expose mjpg/jpg images

### PR DESCRIPTION
**Pull request by jesserockz**
# What does this implement/fix? 

Original PR #2237 (No write access to update that PR)

This uses a `esp32` built-in web-server after the https://github.com/easytarget/esp32-cam-webserver.
This works well for various reasons:

- in general passing `mjpg` stream is bandwidth intensive, so server working in a single threaded mode
   ensures that ESP32 is not overloaded with requests, rather they backlog
- the separate snapshot server is proposed due to above limitation, as: a single stream is allowed, but a snapshot
   is fast to execute so requests for snapshot can be backlogged on a socket
- the separate snapshot server makes it possible to send stream and occasionally from time to time offer a snapshot
   for asking client
- the data are accessed via `GET http://<ip>:<port>/`
- a binary semaphore is used to not use sleep-loop-wait approach
- the `esp32` built-in webserver runs on a `core 0` so no additional synchronization is needed
## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/esp32_camera_web_server/__init__.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+).extend(cv.COMPONENT_SCHEMA)
+
+
+async def to_code(config):
+    server = cg.new_Pvariable(config[CONF_ID])
+    cg.add(server.set_port(config[CONF_PORT]))
+    cg.add(server.set_mode(config[CONF_MODE]))
+    await cg.register_component(server, config)
+
+    cg.add_library("esphome/ESPAsyncWebServer-esphome", "2.0.1")
```
Comment (OttoWinter) [inline esphome/components/esp32_camera_web_server/__init__.py]: wait but this component doesn't use the ESPAsyncWebServer library though, no?

Comment (jesserockz) [reply esphome/components/esp32_camera_web_server/__init__.py]: Haha, yeah I didn't even notice, just fixed an error I had when it conflicted. But just removed this now.

**Comment by ayufan** [comment]
@jesserockz Do you work on this? This is implementation is slightly outdated and does not support all receiving device (due to chunked encoding). Maybe I could integrate your changes into my original PR and merge them with my dev branch https://github.com/ayufan/esphome-components/tree/master/components/esp32_camera_web_server.

**Comment by jesserockz** [comment]
@ayufan yeah I updated with ESPHome Dev etc so I could use it on my devices =)

But yes happy for you to pull my changes back to your original branch and PR.

**Comment by ayufan** [comment]
Hah @jesserockz :) `Dev etc so I could use it on my devices =)` awesome! Happy that this works for you. Then lets finish and merge this into `esphome` so you would not have to do it.

**Pull request closed without merging**